### PR TITLE
[flutter_tools] prevent StateError when log reader is disposed

### LINKED ISSUE #49248: StateError in AdbLogReader when onLine is called after being closed
```
StateError: Bad state: Cannot add new events after calling close

  | at _BroadcastStreamController.add | (broadcast_stream_controller.dart:251 )
-- | -- | --
  | at _AdbLogReader._onLine | (android_device.dart:1115 )
  | at _rootRunUnary | (zone.dart:1134 )
  | at _CustomZone.runUnary | (zone.dart:1031 )
  | at _CustomZone.runUnaryGuarded | (zone.dart:933 )
  | at _BufferingStreamSubscription._sendData | (stream_impl.dart:339 )
  | at _BufferingStreamSubscription._add | (stream_impl.dart:266 )
  | at _SinkTransformerStreamSubscription._add | (stream_transformers.dart:70 )
  | at _EventSinkWrapper.add | (stream_transformers.dart:17 )
  | at _StringAdapterSink.add | (string_conversion.dart:238 )
  | at _LineSplitterSink._addLines | (line_splitter.dart:152 )
  | at _LineSplitterSink.addSlice | (line_splitter.dart:127 )
  | at StringConversionSinkMixin.add | (string_conversion.dart:165 )
  | at _SinkTransformerStreamSubscription._handleData | (stream_transformers.dart:122 )
  | at _rootRunUnary | (zone.dart:1134 )
  | at _CustomZone.runUnary | (zone.dart:1031 )
  | at _CustomZone.runUnaryGuarded | (zone.dart:933 )
  | at _BufferingStreamSubscription._sendData | (stream_impl.dart:339 )
  | at _BufferingStreamSubscription._add | (stream_impl.dart:266 )
  | at _SinkTransformerStreamSubscription._add | (stream_transformers.dart:70 )
  | at _EventSinkWrapper.add | (stream_transformers.dart:17 )
  | at _StringAdapterSink.add | (string_conversion.dart:238 )
  | at _StringAdapterSink.addSlice | (string_conversion.dart:243 )
  | at _Utf8ConversionSink.addSlice | (string_conversion.dart:314 )
  | at _Utf8ConversionSink.add | (string_conversion.dart:307 )
  | at _ConverterStreamEventSink.add | (chunked_conversion.dart:74 )
  | at _SinkTransformerStreamSubscription._handleData | (stream_transformers.dart:122 )
  | at _rootRunUnary | (zone.dart:1134 )
  | at _CustomZone.runUnary | (zone.dart:1031 )
  | at _CustomZone.runUnaryGuarded | (zone.dart:933 )
  | at _BufferingStreamSubscription._sendData | (stream_impl.dart:339 )
  | at _BufferingStreamSubscription._add | (stream_impl.dart:266 )
  | at _SyncStreamControllerDispatch._sendData | (stream_controller.dart:773 )
  | at _StreamController._add | (stream_controller.dart:649 )
  | at _StreamController.add | (stream_controller.dart:595 )
  | at _Socket._onData | (socket_patch.dart:1873 )
  | at _rootRunUnary | (zone.dart:1138 )
  | at _CustomZone.runUnary | (zone.dart:1031 )
  | at _CustomZone.runUnaryGuarded | (zone.dart:933 )
  | at _BufferingStreamSubscription._sendData | (stream_impl.dart:339 )
  | at _BufferingStreamSubscription._add | (stream_impl.dart:266 )
  | at _SyncStreamControllerDispatch._sendData | (stream_controller.dart:773 )
  | at _StreamController._add | (stream_controller.dart:649 )
  | at _StreamController.add | (stream_controller.dart:595 )
  | at new _RawSocket.<anonymous closure> | (socket_patch.dart:1415 )
  | at _NativeSocket.issueReadEvent.issue | (socket_patch.dart:919 )
  | at _microtaskLoop | (schedule_microtask.dart:43 )
  | at _startMicrotaskLoop | (schedule_microtask.dart:52 )
  | at _runPendingImmediateCallback | (isolate_patch.dart:118 )
  | at _RawReceivePortImpl._handleMessage | (isolate_patch.dart:175 )
```

cc @blasten

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Fixes https://github.com/flutter/flutter/issues/49248

Previously the AdbLogReader did async setup in the StreamController.onListen callback, specifically it would query the api version and start the adb process. If the log subscription was cancelled before this setup completed, then the log output *could* (haven't confirmed) get added to a closed controller, causing the above state error. This likely got worse the change to query API version since that added slightly more delay, but in theory it could have always happened.

To fix this, I've hoisted the async setup out of the log reader into a static setup method. This required changing the interface of `Device.getLogReader` to return a `FutureOr<LogReader>`. For safety's sake, I also added a closed check to the line handling method, though I don't think its possible to hit that case anymore, and I've had a heck of a time trying to test it.

At any rate, the AdbLogReader tests have been moved out of android_device_test.dart, and the level of mocking has been reduced significantly as well.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Working on a biggest testing change for this

File: packages/flutter_tools/lib/src/android/android_device.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/android_device.dart]: I think you can override `FutureOr<A> get a` with `A get a`, but I will double check

File: packages/flutter_tools/lib/src/resident_runner.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Same

File: packages/flutter_tools/lib/src/resident_runner.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_runner.dart]: Nope, fixed

File: packages/flutter_tools/lib/src/commands/drive.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/drive.dart]: Done

File: packages/flutter_tools/lib/src/android/android_device.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/android_device.dart]: The following analyzes cleanly so I think this is fine. Also I didn't need to update any of the other Device subclasses:

```

abstract class A {
  FutureOr<int> get foo;
}

abstract class B extends A {
  @override
  int get foo;
}
```

File: packages/flutter_tools/lib/src/android/android_device.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/android_device.dart]: Done

File: packages/flutter_tools/lib/src/android/android_device.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/android_device.dart]: Fixed

File: packages/flutter_tools/lib/src/android/android_device.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/android_device.dart]: Done

**Comment by jonahwilliams** [comment]
I removed the eager initialization in startApp, since I didn't want to go back and update a bunch of test expectations. That would be easier to land once AndroidDevice testing is cleaned up a bit more, because its still kind of a mess

File: packages/flutter_tools/test/src/fake_process_manager.dart
Comment (christopherfujino) [inline packages/flutter_tools/test/src/fake_process_manager.dart]: Why not?

Comment (jonahwilliams) [reply packages/flutter_tools/test/src/fake_process_manager.dart]: This was a bit too opinionated, we kill processes all the time especially if they are servers.

Comment (christopherfujino) [reply packages/flutter_tools/test/src/fake_process_manager.dart]: sgtm

**Comment by christopherfujino** [comment]
LGTM, you should prob wait for @zanderso's final approval.

File: packages/flutter_tools/lib/src/android/android_device.dart
Comment (blasten) [inline packages/flutter_tools/lib/src/android/android_device.dart]: trivial nit: comment about adb exited earlier and it might be still streaming logs.

File: packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart
Comment (blasten) [inline packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart]: minor nit: trailing comma

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart
Comment (blasten) [inline packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart]: nit: trailing comma

File: packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart
Comment (blasten) [inline packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart]: minor nit: `(_) {}`

File: packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart
Comment (blasten) [inline packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart]: minor nit: `group`?

File: packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart]: Done

File: packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart]: Done

File: packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart]: Done

File: packages/flutter_tools/lib/src/android/android_device.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/android/android_device.dart]: Done

File: packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/android/adb_log_reader_test.dart]: I'd rather leave it out if we wouldn't have more than one, since it introduces more rightward drift

**Review by blasten — APPROVED** [review]

**Pull request merged**
